clearcode/store_scans.py

```diff
@@ -0,0 +1,237 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) nexB Inc. and others. All rights reserved.
+#
+# ClearCode is a free software tool from nexB Inc. and others.
+# Visit https://github.com/nexB/clearcode-toolkit/ for support and download.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import defaultdict
+from clearcode.models import CDitem
+from clearcode.cdutils import Coordinate
+from clearcode.cdutils import str2coord
+from django.db.models import Q
+from hashlib import sha512
+import json
+import requests
+from packageurl import PackageURL
+from pathlib import Path
+from git import Repo
+import os
+"""
+The input is a bunch of scans from ClearlyDefined and 
+the output is a bunch of git repositories with commited and 
+pushed scans such that we balance the scans roughly evenly accross 
+different repositories.
+
+The primary reason for multiple repositories is size of a single 
+repo. There is a size limit of 5 GB at GitHub and it's difficult 
+to work with repositories with million files.
+
+Therefore the approach is to use hashing as a way to name git 
+repositories and directories. We compute hash on the purl of the scanned 
+package and use the first few layers of this hash for the repo and 
+directory names.
+
+Initial processing steps are:
+- We collect a list of scan and purl.
+- For each we compute a hash and determine the repo and directory.
+- If needed we create the repo or pull it.
+- Then we store the scan using the purl hash and purl as path.
+- Finally commit and push! : )
+
+Because it's not practical to process many repos at once, we organize the 
+processing one repo a time. For this, we iterate over a bunch of records get or compute 
+the purl hash and process the records that share the same hash.
+
+We are using a short hash that is three characters long using hexadecimal encoding. 
+Therefore we can have 16*16*16 = 4096 repositories where each repo would contain about 
+25k scan files, if we were to store 100 million scans (which is a high mark).
+For reference one scan should use less than a 100k on average when compressed 
+with gzip or git based on looking at 15 million scans. Each repo should be roughly 
+couple hundred mega bytes big, based on 15 million scans.
+"""
+
+repo_names = [hex(hash)[2:].zfill(3) for hash in range(4096)]
```

Comment:
Can you add a comment about what is going on in this list comprehension? I see that you are creating hex values of integers and then ignoring the 0x prefix, but it is not obvious at first glance.